File: backend/internal/service/orders_25.go
package service


import (
	"context"
	"time"

	"github.com/biruk-ak/clinical-lis/backend/internal/models"
	"github.com/biruk-ak/clinical-lis/backend/internal/repository"
	"github.com/google/uuid"
)

// OrderRecord25Service encapsulates business rules for OrderRecord25.
type OrderRecord25Service struct {
	repo *repository.OrderRecord25Repository
}

// NewOrderRecord25Service wires the service.
func NewOrderRecord25Service(repo *repository.OrderRecord25Repository) *OrderRecord25Service {
	return &OrderRecord25Service{repo: repo}
}

// Create validates and persists a new OrderRecord25.
func (s *OrderRecord25Service) Create(ctx context.Context, actor, facilityID, code, display, status string, priority int, notes string) (*models.OrderRecord25, error) {
	now := time.Now().UTC()
	rec := &models.OrderRecord25{
		ID:           uuid.NewString(),
		ExternalCode: code,
		DisplayName:  display,
		Status:       status,
		Priority:     priority,
		FacilityID:   facilityID,
		CreatedBy:    actor,
		UpdatedBy:    actor,
		Notes:        notes,
		MetadataJSON: "{}",
		Version:      1,
		IsActive:     true,
		CreatedAt:    now,
		UpdatedAt:    now,
	}
	if err := rec.Validate(); err != nil {
		return nil, err
	}
	if err := s.repo.Create(ctx, rec); err != nil {
		return nil, err
	}
	return rec, nil
}

// Get returns a OrderRecord25 by id.
func (s *OrderRecord25Service) Get(ctx context.Context, id string) (*models.OrderRecord25, error) {
	return s.repo.GetByID(ctx, id)
}

// List returns facility-scoped OrderRecord25 pages.
func (s *OrderRecord25Service) List(ctx context.Context, facilityID string, limit, offset int) ([]models.OrderRecord25, error) {
	if limit <= 0 || limit > 200 {
		limit = 50
	}
	if offset < 0 {
		offset = 0
	}
	return s.repo.List(ctx, facilityID, limit, offset)
}

// Update mutates an existing OrderRecord25.
func (s *OrderRecord25Service) Update(ctx context.Context, actor string, rec *models.OrderRecord25) error {
	if err := rec.Validate(); err != nil {
		return err
	}
	rec.Touch(actor, time.Now().UTC())
	return s.repo.Update(ctx, rec)
}

// Archive soft-deletes a OrderRecord25.
func (s *OrderRecord25Service) Archive(ctx context.Context, id, actor string) error {
	return s.repo.Delete(ctx, id, actor)
}

// Search finds OrderRecord25 by external code prefix.
func (s *OrderRecord25Service) Search(ctx context.Context, facilityID, prefix string) ([]models.OrderRecord25, error) {
	return s.repo.SearchByCode(ctx, facilityID, prefix, 100)
}

// Stats returns active counts for dashboards.
func (s *OrderRecord25Service) Stats(ctx context.Context, facilityID string) (int64, error) {
	return s.repo.CountActive(ctx, facilityID)
}

File: frontend/src/types/orders25.ts
/** Domain types for orders / OrderRecord25 */
export type OrderRecord25Status = 'draft' | 'active' | 'pending' | 'completed' | 'cancelled' | 'archived';

export interface OrderRecord25 {
  id: string;
  externalCode: string;
  displayName: string;
  status: OrderRecord25Status;
  priority: number;
  facilityId: string;
  createdBy: string;
  updatedBy: string;
  notes: string;
  metadataJson: string;
  version: number;
  isActive: boolean;
  createdAt: string;
  updatedAt: string;
  archivedAt?: string | null;
}

export interface OrderRecord25CreateRequest {
  externalCode: string;
  displayName: string;
  status: OrderRecord25Status;
  priority: number;
  facilityId: string;
  notes?: string;
}

export interface OrderRecord25ListResponse {
  items: OrderRecord25[];
  total?: number;
}

export interface OrderRecord25StatsResponse {
  activeCount: number;
}

export function isOrderRecord25Active(rec: OrderRecord25): boolean {
  return rec.isActive && rec.status !== 'archived' && rec.status !== 'cancelled';
}

export function formatOrderRecord25Label(rec: OrderRecord25): string {
  return `${rec.externalCode} — ${rec.displayName}`;
}

export function compareOrderRecord25ByPriority(a: OrderRecord25, b: OrderRecord25): number {
  if (a.priority !== b.priority) return b.priority - a.priority;
  return a.displayName.localeCompare(b.displayName);
}
